fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/clientlog/KeySpaceCountTree.java

```diff
@@ -83,35 +389,181 @@ protected TupleKeyCountTree newPrefixChild(@Nonnull Object prefix) {
+                            break;
+                        }
+                        return CompletableFuture.completedFuture(new ResolvedRecordTypeKeyspace(resolvedParent, recordType));
+                    }
+                    KeyExpression commonPrimaryKey = recordStoreKeyspace.getRecordMetaData().commonPrimaryKey();
+                    if (commonPrimaryKey != null && distance < commonPrimaryKey.getColumnSize()) {
+                        return resolvePrimaryKeyField(context, resolvedParent, object, commonPrimaryKey, distance);
+                    }
+                    break;
+                case INDEX:
+                case INDEX_SECONDARY_SPACE:
+                case INDEX_RANGE_SPACE:
+                case INDEX_UNIQUENESS_VIOLATIONS_SPACE:
+                case INDEX_BUILD_SPACE:
+                    if (distance == 0 && object != null) {
+                        final Index index;
+                        try {
+                            index = recordStoreKeyspace.getRecordMetaData().getIndexFromSubspaceKey(object);
+                        } catch (RecordCoreException ex) {
+                            break;
+                        }
+                        return CompletableFuture.completedFuture(new ResolvedIndexKeyspace(resolvedParent, index));
+                    }
+                    break;
+                default:
+                    break;
+            }
+        }
+        if (recordTypeKeyspace != null) {
+            KeyExpression primaryKey = recordTypeKeyspace.getRecordType().getPrimaryKey();
+            if (distance + 1 < primaryKey.getColumnSize()) {
+                return resolvePrimaryKeyField(context, resolvedParent, object, primaryKey, distance + 1);
+            }
         }
+        if (indexKeyspace != null &&
+                indexKeyspace.getParent() instanceof ResolvedRecordStoreKeyspace &&
+                ((ResolvedRecordStoreKeyspace)indexKeyspace.getParent()).getRecordStoreKeyspace() == FDBRecordStoreKeyspace.INDEX) {
+            Index index = indexKeyspace.getIndex();
+            KeyExpression indexExpression = index.getRootExpression();
+            if (distance < indexExpression.getColumnSize()) {
+                return resolveIndexField(context, resolvedParent, object, index, indexExpression, distance);
+            }
+        }
+        return UNRESOLVED;
+    }
+
+    /**
+     * Determine whether this leaf of the {@link KeySpaceDirectory} tree is the root of a {@link com.apple.foundationdb.record.provider.foundationdb.FDBRecordStore}.
+     * @param context an open transaction to use to read from the database
+     * @param resolvedParent the resolved parent node
+     * @param object the {@link com.apple.foundationdb.tuple.Tuple} element for this node
+     * @return {@code true} if this path stores a record store
+     */
+    protected boolean isRecordStoreLeaf(@Nonnull FDBRecordContext context, @Nonnull Resolved resolvedParent, @Nullable Object object) {
```

Comment:
Maybe not for this PR, but it seems that it may be very useful to formalize where record stores live in the keyspace (so that a directory "knows" that is contains record stores)...but, I guess for this to really be useful, then the directory should also have some way to describe how to find the metadata for the stores under it.  Anyway, just thinking out loud.